src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp

```diff
@@ -183,6 +183,29 @@ void ShenandoahGenerationalHeap::stop() {
   regulator_thread()->stop();
 }
 
+bool ShenandoahGenerationalHeap::requires_barriers(stackChunkOop obj) const {
+  if (is_idle()) {
+    return false;
+  }
+
+  if (is_concurrent_young_mark_in_progress() && is_in_young(obj) && !marking_context()->allocated_after_mark_start(obj)) {
```

Comment:
Just so I understand (and maybe this deserves a clarifying comment to help others understand):
  The barrier we are concerned about here is SATB.  If obj is above TAMS, there's no need to enforce a SATB barrier.  Any reference residing above TAMS which is overwritten will have been marked already, or is about to be marked.  Either the value is already "presumed marked" because it also resides above TAMS, or its value was assigned from a root (in which case root scanning marked it), or it was copied from a pre-existing object (in which case it will be marked by scanning of that pre-existing object or by SATB).

Reply:
Yes, concerned about SATB here.